### PR TITLE
ZEPPELIN-3569. Improvement of FlinkInterpreter

### PR DESCRIPTION
**Pull request by zjffdu**
### What is this PR for?
This PR just refactor the flink interpreter and also introduce several main features. 

Here's the main changes;

1. Upgrade flink to 1.5.2
2. Support ZeppelinContext
3. Support %flink.sql
4. Support yarn mode

### What type of PR is it?
[Improvement | Feature]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-3569

### How should this be tested?
* Unit test is added

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by zjffdu** [comment]
There's one critical issue in flink 1.5.0 FLINK-9554

File: bin/interpreter.sh [hunk 1]
```diff
@@ -204,6 +204,21 @@ elif [[ "${INTERPRETER_ID}" == "pig" ]]; then
   else
     echo "TEZ_CONF_DIR is not set, configuration might not be loaded"
   fi
+elif [[ "${INTERPRETER_ID}" == "flink" ]]; then
+  if [[ -n "${HADOOP_CONF_DIR}" ]] && [[ -d "${HADOOP_CONF_DIR}" ]]; then
+    ZEPPELIN_INTP_CLASSPATH+=":${HADOOP_CONF_DIR}"
+    export HADOOP_CONF_DIR=${HADOOP_CONF_DIR}
```
Comment (felixcheung) [inline bin/interpreter.sh]: why do we need this if it's already set?

File: bin/interpreter.sh [hunk 2]
```diff
@@ -204,6 +204,21 @@ elif [[ "${INTERPRETER_ID}" == "pig" ]]; then
...
+  else
+    # autodetect HADOOP_CONF_HOME by heuristic
+    if [[ -n "${HADOOP_HOME}" ]] && [[ -z "${HADOOP_CONF_DIR}" ]]; then
+      if [[ -d "${HADOOP_HOME}/etc/hadoop" ]]; then
+        export HADOOP_CONF_DIR="${HADOOP_HOME}/etc/hadoop"
+      elif [[ -d "/etc/hadoop/conf" ]]; then
```
Comment (felixcheung) [inline bin/interpreter.sh]: this else should be also when HADOOP_HOME is not set

File: flink/src/main/java/org/apache/zeppelin/flink/FlinkInterpreter.java [hunk 3]
```diff
@@ -6,412 +6,94 @@
  * to you under the Apache License, Version 2.0 (the
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
```
Comment (felixcheung) [inline flink/src/main/java/org/apache/zeppelin/flink/FlinkInterpreter.java]: not sure why we have this?

File: flink/src/main/scala/org/apache/zeppelin/flink/FlinkScalaInterpreter.scala [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    val in0 = getField(flinkILoop, "scala$tools$nsc$interpreter$ILoop$$in0").asInstanceOf[Option[BufferedReader]]
+    val reader = in0.fold(flinkILoop.chooseReader(settings))(r => SimpleReader(r, replOut, interactive = true))
+
+    flinkILoop.in = reader
+    flinkILoop.initializeSynchronous()
+    callMethod(flinkILoop, "scala$tools$nsc$interpreter$ILoop$$loopPostInit")
+    this.scalaCompleter = reader.completion.completer()
+
+    this.benv = flinkILoop.scalaBenv
+    this.senv = flinkILoop.scalaSenv
```
Comment (felixcheung) [inline flink/src/main/scala/org/apache/zeppelin/flink/FlinkScalaInterpreter.scala]: do we need to bind on this?

File: bin/interpreter.sh (hunk 1 above)
Comment (zjffdu) [reply bin/interpreter.sh]: flink yarn mode needs `HADOOP_CONF_DIR`

File: flink/src/main/scala/org/apache/zeppelin/flink/FlinkScalaInterpreter.scala (hunk 4 above)
Comment (zjffdu) [reply flink/src/main/scala/org/apache/zeppelin/flink/FlinkScalaInterpreter.scala]: `flinkILoop` already did the binding.

File: bin/interpreter.sh (hunk 1 above)
Comment (felixcheung) [reply bin/interpreter.sh]: `if [[ -n "${HADOOP_CONF_DIR}" ]]` this says `HADOOP_CONF_DIR` is already in the environment?

Comment (zjffdu) [reply bin/interpreter.sh]: yes

File: bin/interpreter.sh (hunk 2 above)
Comment (zjffdu) [reply bin/interpreter.sh]: if condition already check `${HADOOP_HOME}/etc/hadoop`

**Comment by felixcheung** [comment]
this comment? https://github.com/apache/zeppelin/pull/3054#discussion_r200825969
it's in every file touched

File: bin/interpreter.sh (hunk 1 above)
Comment (felixcheung) [reply bin/interpreter.sh]: so then why export the exact same thing again?

File: flink/src/main/java/org/apache/zeppelin/flink/FlinkInterpreter.java (hunk 3 above)
Comment (zjffdu) [reply flink/src/main/java/org/apache/zeppelin/flink/FlinkInterpreter.java]: Fixed

**Pull request closed without merging**